src/contracts/Qearn.h

```diff
@@ -0,0 +1,649 @@
+        REGISTER_USER_PROCEDURE(Lock, 1);
+		REGISTER_USER_PROCEDURE(Unlock, 2);
+
+	_
+
+    struct BEGIN_EPOCH_locals
+    {
+        HistoryInfo INITIALIZE_HISTORY;
+        UserInfo INITIALIZE_USER;
+        RoundInfo INITIALIZE_ROUNDINFO;
+
+        uint32 t;
+        bit status;
+        uint64 pre_epoch_balance;
+        uint64 current_balance;
+        ::Entity entity;
+    };
+
+    BEGIN_EPOCH_WITH_LOCALS
+
+        qpi.getEntity(SELF, locals.entity);
+        locals.current_balance = locals.entity.incomingAmount - locals.entity.outgoingAmount;
+
+        locals.pre_epoch_balance = 0UL;
+        for(locals.t = qpi.epoch() - 1; locals.t >= qpi.epoch() - 52; locals.t--) 
+        {
+            locals.pre_epoch_balance += state._CurrentRoundInfo.get(locals.t)._Epoch_Bonus_Amount + state._CurrentRoundInfo.get(locals.t)._Total_Locked_Amount;
+        }
+
+        locals.INITIALIZE_ROUNDINFO._Epoch_Bonus_Amount = locals.current_balance - locals.pre_epoch_balance;
+        locals.INITIALIZE_ROUNDINFO._Total_Locked_Amount = 0;
+
+        state._InitialRoundInfo.set(qpi.epoch(), locals.INITIALIZE_ROUNDINFO);
+        state._CurrentRoundInfo.set(qpi.epoch(), locals.INITIALIZE_ROUNDINFO);
+
+        if(locals.INITIALIZE_ROUNDINFO._Epoch_Bonus_Amount > 0) 
+        {
+            
+            locals.INITIALIZE_ROUNDINFO._Epoch_Bonus_Amount += state.remain_amount;
+
+            state._InitialRoundInfo.set(qpi.epoch(), locals.INITIALIZE_ROUNDINFO);
+            state._CurrentRoundInfo.set(qpi.epoch(), locals.INITIALIZE_ROUNDINFO);
+
+            state.remain_amount = 0;
+
+        }
+	_
+
+    struct END_EPOCH_locals 
+    {
+        HistoryInfo INITIALIZE_HISTORY;
+        UserInfo INITIALIZE_USER;
+        RoundInfo INITIALIZE_ROUNDINFO;
+
+        uint64 _reward_percent;
+        uint64 _reward_amount;
+        uint64 _burn_amount;
+        uint32 _count_Of_last_vacancy;
+        uint32 _t;
+
+    };
+
+	END_EPOCH_WITH_LOCALS
+
+        locals._count_Of_last_vacancy = 0;
+        state._Unlocked_cnt = 0;
+        
+        locals._burn_amount = state._CurrentRoundInfo.get(qpi.epoch() - 52)._Epoch_Bonus_Amount;
+        
+        if(state._CurrentRoundInfo.get(qpi.epoch() - 52)._Total_Locked_Amount > 0) 
+        {
+            locals._reward_percent = state._CurrentRoundInfo.get(qpi.epoch() - 52)._Epoch_Bonus_Amount * 10000000UL / state._CurrentRoundInfo.get(qpi.epoch() - 52)._Total_Locked_Amount;
+        }
+
+        for(locals._t = 0 ; locals._t < state._Locked_Last_Index; locals._t++) 
+        {
+            if(state.Locker.get(locals._t)._Locked_Epoch == qpi.epoch() - 52) 
+            {
+
+                if(state.Locker.get(locals._t)._Locked_Amount == 0) 
+                { //  the case to be unlocked early
+
+                    locals._count_Of_last_vacancy++; continue;
+                    
+                }
+
+                locals._reward_amount = state.Locker.get(locals._t)._Locked_Amount * locals._reward_percent / 10000000UL;
```

Comment:
please don't use `/`, use qpi::div instead
apply this for all `/` in this file

Reply:
fixed